### PR TITLE
Bindings-3.5.5.2 Check url against destination

### PR DESCRIPTION
**Pull request by bakejeyner**
Adds Binding verification for destination checking.

### CODE REVIEW

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt [hunk 1]
```diff
@@ -17,12 +17,14 @@ import io.kotlintest.matchers.shouldNotBe
...
 import org.codice.compliance.SAMLComplianceException
 import org.codice.compliance.SAMLSpecRefMessage.*
+import org.codice.compliance.allChildren
 import org.codice.compliance.children
 import org.codice.compliance.utils.TestCommon
 import org.codice.compliance.utils.TestCommon.Companion.EXAMPLE_RELAY_STATE
 import org.codice.compliance.utils.TestCommon.Companion.MAX_RELAYSTATE_LEN
 import org.codice.compliance.utils.decorators.IdpPostResponseDecorator
 import org.codice.security.sign.Decoder
+import org.w3c.dom.Node
```
Comment (coyotesqrl) [inline test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt]: ❗️ Unused import

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt [hunk 2]
```diff
@@ -16,7 +16,9 @@ package org.codice.compliance.verification.binding
 import org.apache.cxf.rs.security.saml.sso.SSOConstants.SAML_RESPONSE
 import org.codice.compliance.SAMLComplianceException
 import org.codice.compliance.SAMLSpecRefMessage.*
+import org.codice.compliance.allChildren
 import org.codice.compliance.children
+import org.codice.compliance.utils.TestCommon
 import org.codice.compliance.utils.TestCommon.Companion.ACS_URL
```
Comment (coyotesqrl) [inline test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt]: ❗️ Unused import.

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt [hunk 3]
```diff
@@ -26,6 +28,7 @@ import org.codice.security.sign.Decoder
 import org.codice.security.sign.Decoder.DecoderException.InflErrorCode.*
 import org.codice.security.sign.SimpleSign
 import org.codice.security.sign.SimpleSign.SignatureException.SigErrorCode
+import org.w3c.dom.Node
```
Comment (coyotesqrl) [inline test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt]: ❗️ Unused import.

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt [hunk 4]
```diff
@@ -17,12 +17,14 @@ import io.kotlintest.matchers.shouldNotBe
...
 
 class PostBindingVerifier(val response: IdpPostResponseDecorator) {
```
Comment (coyotesqrl) [inline test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt]: ❗️ `response` can be private. The same applies in `RedirectBindingVerifier`.

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt [hunk 5]
```diff
@@ -94,13 +96,18 @@ class PostBindingVerifier(val response: IdpPostResponseDecorator) {
...
+     * Verifies the destination is correct according to the post binding rules in the binding spec
+     * 3.5.5.2 Security Considerations
      */
     private fun verifyPostDestination() {
         val destination = response.responseDom.attributes.getNamedItem("Destination")?.nodeValue
-        if (destination != TestCommon.ACS_URL) {
-            throw SAMLComplianceException.createWithPropertyNotEqualMessage(SAMLBindings_3_4_5_2_a1,
+        val signatures = response.responseDom.allChildren("Signature")
+
+        // "If the message is signed, the Destination XML attribute in the root SAML element of the
```
Comment (coyotesqrl) [inline test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt]: ✏️ I'm not sure I see a ton of value putting a comment here that has the same text as the message to be returned.

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt [hunk 6]
```diff
@@ -94,13 +96,18 @@ class PostBindingVerifier(val response: IdpPostResponseDecorator) {
...
+        // protocol message MUST contain the URL to which the sender has instructed the user agent
+        // to deliver the message"
+        if ( signatures.isNotEmpty() && destination != TestCommon.ACS_URL) {
```
Comment (coyotesqrl) [inline test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt]: ❗️ Reformat.

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt [hunk 7]
```diff
@@ -200,11 +203,16 @@ class RedirectBindingVerifier(val response: IdpRedirectResponseDecorator) {
...
     private fun verifyRedirectDestination() {
         val destination = response.responseDom.attributes.getNamedItem("Destination")?.nodeValue
-        if (destination != ACS_URL) {
-            throw SAMLComplianceException.createWithPropertyNotEqualMessage(SAMLBindings_3_4_5_2_a1,
+        val signatures = response.responseDom.allChildren("Signature")
+
+        // "If the message is signed, the Destination XML attribute in the root SAML element of the
+        // protocol message MUST contain the URL to which the sender has instructed the user agent
+        // to deliver the message"
+        if ( signatures.isNotEmpty() && destination != TestCommon.ACS_URL) {
```
Comment (coyotesqrl) [inline test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt]: ❗️ Reformat.

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt [hunk 8]
```diff
@@ -200,11 +203,16 @@ class RedirectBindingVerifier(val response: IdpRedirectResponseDecorator) {
      */
     private fun verifyRedirectDestination() {
         val destination = response.responseDom.attributes.getNamedItem("Destination")?.nodeValue
-        if (destination != ACS_URL) {
-            throw SAMLComplianceException.createWithPropertyNotEqualMessage(SAMLBindings_3_4_5_2_a1,
+        val signatures = response.responseDom.allChildren("Signature")
+
+        // "If the message is signed, the Destination XML attribute in the root SAML element of the
```
Comment (coyotesqrl) [inline test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt]: ✏️ Same comment about the comment text matching the error.

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt [hunk 9]
```diff
@@ -200,11 +203,16 @@ class RedirectBindingVerifier(val response: IdpRedirectResponseDecorator) {
...
+        // protocol message MUST contain the URL to which the sender has instructed the user agent
+        // to deliver the message"
+        if ( signatures.isNotEmpty() && destination != TestCommon.ACS_URL) {
+            throw SAMLComplianceException.createWithPropertyNotEqualMessage(SAMLBindings_3_5_5_2_a,
                     "Destination",
                     destination,
-                    ACS_URL)
+                    TestCommon.ACS_URL)
```
Comment (brjeter) [inline test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt]: ❓ Is the relevant section of spec here duplicated in both redirect and posting bindings? I'm concerned that this method is identical to the PostBinding one.

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt (hunk 5 above)
Comment (brjeter) [reply test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt]: Agreed

Comment (bakejeyner) [reply test/common/src/main/kotlin/org/codice/compliance/verification/binding/PostBindingVerifier.kt]: 👍

File: test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt (hunk 9 above)
Comment (bakejeyner) [reply test/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt]: Yes. However I feel like we shouldn't refactor until more of the verification is finished and we can see how much duplication we will have and if we want to refactor at all.

**Pull request merged**